net/httpclient.go

```diff
@@ -70,10 +73,14 @@ func NewClient(o Options) *Client {
 }
 
 func (c *Client) Close() {
-	c.tr.Close()
-	if c.sr != nil {
-		c.sr.Close()
-	}
+	c.once.Do(func() {
+		if c.tr != nil {
+			c.tr.Close()
+		}
+		if c.sr != nil {
+			c.sr.Close()
+		}
```

Comment:
Why do we need these nil checks?

Reply:
yes otherwise:
```
% go test ./net                                                                 test/go-leak
DO NOT USE IN PRODUCTION
--- FAIL: TestClient (0.00s)
    --- FAIL: TestClient/All_defaults,_with_request_should_have_a_response (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xabe033]

goroutine 123 [running]:
panic({0xcb2160, 0x14543b0})
        /usr/share/go/src/runtime/panic.go:987 +0x3ba fp=0xc00001d7b8 sp=0xc00001d6f8 pc=0x4398da
testing.tRunner.func1.2({0xcb2160, 0x14543b0})
        /usr/share/go/src/testing/testing.go:1396 +0x24e fp=0xc00001d868 sp=0xc00001d7b8 pc=0x5203ae
testing.tRunner.func1()
        /usr/share/go/src/testing/testing.go:1399 +0x39f fp=0xc00001d9d0 sp=0xc00001d868 pc=0x51fe3f
runtime.deferCallSave(0xc00001daa0, 0xc00001df90?)
        /usr/share/go/src/runtime/panic.go:796 +0x88 fp=0xc00001d9e0 sp=0xc00001d9d0 pc=0x4394c8
runtime.runOpenDeferFrame(0xc000482190?, 0xc000482140)
        /usr/share/go/src/runtime/panic.go:769 +0x1a5 fp=0xc00001da28 sp=0xc00001d9e0 pc=0x4392e5
panic({0xcb2160, 0x14543b0})
        /usr/share/go/src/runtime/panic.go:884 +0x212 fp=0xc00001dae8 sp=0xc00001da28 pc=0x439732
runtime.panicmem(...)
        /usr/share/go/src/runtime/panic.go:260
runtime.sigpanic()
        /usr/share/go/src/runtime/signal_unix.go:835 +0x2f6 fp=0xc00001db38 sp=0xc00001dae8 pc=0x450716
github.com/zalando/skipper/net.(*Client).Close.func1()
        /home/sszuecs/go/src/github.com/zalando/skipper/net/httpclient.go:81 +0x33 fp=0xc00001db58 sp=0xc00001db38 pc=0xabe033
sync.(*Once).doSlow(0xdd1d90?, 0x21?)
        /usr/share/go/src/sync/once.go:74 +0xc2 fp=0xc00001dbb8 sp=0xc00001db58 pc=0x4783c2
sync.(*Once).Do(...)
        /usr/share/go/src/sync/once.go:65
github.com/zalando/skipper/net.(*Client).Close(0xc000134600?)
        /home/sszuecs/go/src/github.com/zalando/skipper/net/httpclient.go:79 +0x51 fp=0xc00001dbe8 sp=0xc00001dbb8 pc=0xabdfd1
github.com/zalando/skipper/net.TestClient.func1.4()
        /home/sszuecs/go/src/github.com/zalando/skipper/net/httpclient_test.go:164 +0x26 fp=0xc00001dc00 sp=0xc00001dbe8 pc=0xac8146
github.com/zalando/skipper/net.TestClient.func1(0xc0002031e0)
        /home/sszuecs/go/src/github.com/zalando/skipper/net/httpclient_test.go:199 +0xca4 fp=0xc00001df70 sp=0xc00001dc00 pc=0xac8084
testing.tRunner(0xc0002031e0, 0xc000012318)
```
which is a bit unclear how this can be true for `tr/sr` ...
These seem to be safe during initialization.

Reply:
Check out `shallowCopy`, I think it is missing `sr` copy. Maybe we can use pointer to `sync.Once` to use `tt := *t` and re-initialize it inside of `shallowCopy`.
TBH I do not thin that `WithX` API that requires `shallowCopy` is good

Reply:
I think ptr to sync.Once is not a good idea. You might want to close the original after you got your replacement and then you closed also your replacement....

Reply:
> and then you closed also your replacement

No, I mean
```go
func (t *Transport) shallowCopy() *Transport {
	tt := *t
	tt.once = &sync.Once{}
	return &tt
```
to automatically ensure all other fields are copied

Reply:
`sr` is not part of the Transport, but the Client so I think also the mentioned change won't be correct.

Reply:
Ah, ok, `tr` can not be nil but `sr` can.

The mentioned change is nice to ensure we do not forget to copy client fields we add later

Reply:
ok removed nil check for transport